### ISSUE TITLE
webpack2 doesn't support IE8

### ISSUE THREAD

**Issue by breeswish** [webpack/webpack#3070]
es5-shim cannot polyfill getters or setters of Object.defineProperty:

https://github.com/webpack/webpack/blob/master/buildin/module.js#L10

**Comment by jdalton** [comment]
jQuery, Prototype, React, & h5bp have already dropped IE8 support and Microsoft only really supports IE 11 and Edge now.

I think dropping IE8 support is a fine thing for Webpack 2. With that Webpack 2 can update it's [buffer](https://www.npmjs.com/package/buffer) dependency to v5.

**Comment by breeswish** [comment]
ok here is my idea. I think at least the most commonly used part (in this case, the module) should be compatible with ALL JavaScript
 enabled, a-well-accepted-standard implemented browsers.

I don't care about buffer. Those things come from Nodejs, which is not expected to work in browser before webpack. Webpack make it work in browsers. Good. with a browser version requirement. Accepted. 

Webpack is only a module bundler, acts as a build tool chain as well nowadays. IT SHOULD NOT MAKE WORKING THINGS DOESN'T WORK ANY LONGER IN THE SAME ENVIRONMENT. I think this is as unreasonable as gulp or grunt doesn't support ES3.

In addition, browserify supports IE8 and it's core functionality (I mean, without models polyfills) can even work in IE6.

**Comment by jdalton** [comment]
IE8 is a dead browser on a dead OS. At some point you have to move on.

**Comment by breeswish** [comment]
Thanks I agree with that. It is dead. It should be replaced. However I would like to express my thoughts: what webpack should do is to bundle modules. what it shouldn't do is to make working things not working. Even babel won't break existing working code.

**Comment by jdalton** [comment]
> Even babel won't break existing working code.

Major bumps avoid breaking working code as long as your deps ranges are following semver. Babel  has introduced several breaking changes on their major version bumps *(that's the primary reason for a major bump)*

**Comment by breeswish** [comment]
Sorry I'm not expressing clearly. I mean, by design, a toolchain which act as a transformer should not break working things.

**Comment by jdalton** [comment]
Babel is really a poor example. Babel v6 `preset-es2015` preset lacks IE8 support out of the box and will generate code that doesn't run in IE8.

**Comment by jdalton** [comment]
Why not stick with Webpack v1 for unsupported browsers?

**Comment by breeswish** [comment]
Yes that's why I am using the term "some". ES modules won't work in ES3, out-of-the-box when using `es2015` preset. However by using other plugins it can be achieved. Can webpack 2 supports ES3 by using other plugins? No. Babel itself doesn't introduce side effects. Webpack 2 itself introduce unavoidable side effects. That's a design issue.

**Comment by jdalton** [comment]
> That's a design issue.

Naw. Putting the burden on Webpack to support dead environments is a design issue. Heck you could always run a post process on your code to replace the bits IE8 has issues with. Trying to push your niche concern onto Webpack is pretty lame.

**Comment by breeswish** [comment]
@jdalton Sorry but in fact it is really hard to run a post-processor. As said before, only some of the ES6 feature can be transformed to be compatible in ES3, which means for the rest of the features it is difficult. Unfortunately, getters and setters are one of those hard to post-process: https://github.com/webpack/webpack/blob/master/buildin/module.js#L10

**Comment by breeswish** [comment]
In fact according to NetApplications or StatCounter (the two most authoritative stat), IE8 is the second most popular browser in IE family and has a market share larger than 2%. In NetApplications, IE 8 is even the 3rd popular browser in global and has a market share 6.25% in the last year. In my own opinion the data of NetApplications should be more accurate because StatCounter doesn't take Internet blocking (for example China) into account and the samples of China is significant (700 million, 22% of the global). Especially in China, IE8 has a market share more than 16%, which is the second most popular browser, according to China's largest search engine Baidu (notice that Google is not accessible in China).

In short, IE 8 is **not a dead environment**, especially in some country.

> I see putting the burden on Webpack v2 to support dead environments as an issue

Let me explain my core idea: webpack is not a library. It is a toolchain. A toolchain should introduce minimum side effects to projects using it. That's it. It should not make working code not working, especially due to some completely avoidable issues.

**Comment by jdalton** [comment]
@breeswish 
> Sorry but in fact it is really hard to run a post-processor.

No it's not. You write a simple Node script to load your file in and regex parse out the bits.
Get has hacky as you want. 

> In fact according to NetApplications or StatCounter

You're cherry picking a bit. The [h5bp thread](https://github.com/h5bp/html5-boilerplate/issues/1524#issuecomment-238868457) I linked to showed that most stats report it low:
<img width="674" src="https://cloud.githubusercontent.com/assets/4303/18917932/00a6d7de-854e-11e6-87f5-4b686b4e4f47.png">

> In short, IE 8 is not a dead environment, especially in some countries, for example, China, which has the MOST netizens in the world.

I believe even in China there's access to other browsers. I recommend [directing your users to alternative browsers](http://browsehappy.com/). Folks still on IE8 are used to an increasingly degraded experience. There is no place in today's toolchain for an 8 year old browser.
I'm sorry you're mildly inconvenienced to support it.

> The idea of only supporting environments you willing to, is applicable for a toy project. It is not applicable for an engineering project.

You're free to use whatever you want including Webpack v1. As I said jQuery, Lodash, Ember, Angular, Prototype, React, & h5bp have already dropped IE8 support . Even Uglify-JS defaults to "screw-ie8" enabled. The burden of support for dead environments is on you.

**Comment by sokra** [comment]
If this code fragment is the only issue for ie8, feel free to find the commit changing loaded to l and id to i an revert it in a PR.

But keep in mind that the whole es modules stuff is based on getter an won't run on ie8. CommonJs may work.

**Comment by jdalton** [comment]
@sokra You're up for [larger](https://github.com/webpack/webpack/commit/902220a0ea51cbd1c117ff274238eb184131b99f) bundles to support IE8?

**Comment by breeswish** [comment]
@jdalton I prefer to provide an opt-out for a fallback, like `loose` mode in babel or other modules. Besides, the size advantage by replacing longer words into shorter ones is minimum when gzip is applied.

**Comment by jhnns** [comment]
Sorry, @sokra but I don't think we should do that.

It *may* be just this little code line for now, but will future changes also be tested on IE8? If we do support IE8, we must ensure that all future changes also do support IE8. That's where I agree with @jdalton: A major version bump is the time to move on.

Shipping code for IE8 is **irresponsibly** because there are no security fixes for this browser + OS anymore. The web **should** look broken for them.

**Comment by lilijialiang** [comment]
# webpack2 doesn't support IE8 !

**Comment by tpizza** [comment]
While I agree for the most part in regards to not supporting IE8 there is still a considerable subset of web applications which still have to support IE8.  It sucks, I know first-hand, heh.

I mean, sure I can hang back on v1.13, but ideally it would be nice if output at least remained functional...or at least offered some sort of configurable settings for enabling ES3 compliant output.  If this does not happen, I totally get it since IE8 support is such a costly endeavor, but it would be nice to be able to continue to drive Webpack adoption across some of my other teams who do not have a choice but to continue supporting that ugly abandoned browser.  I've been pushing to try and sunset IE8 completely and at least bump to 9+, but in the B2B world...that burn can be too slow!

**Comment by jdalton** [comment]
> I mean, sure I can hang back on v1.13,

I'm glad you see that as an option because doors everywhere are closing on IE8.

**Comment by sokra** [comment]
After some discussion we decided for the following statement regarding IE8:

We don't support IE8 in webpack 2. IE8 is not supported by Microsoft and doesn't get security fixes. We think it's no responsible to allow web developers to build webapps which support IE8. They must break!

**Comment by drewfreyling** [comment]
"With that Webpack 2 can update its buffer dependency to v5 too!"
@jdalton - to be clear that would remove IE9 and IE10 support too (Which [browserify](https://github.com/substack/node-browserify/blob/master/changelog.markdown#1400) has done as of v14).

**Comment by csvan** [comment]
Rejecting this was a great move. Any effort to enable IE8 to keep existing is utterly irresponsible and ultimately detrimental to developers and users alike.

**Comment by ericyang89** [comment]
i meet the same issue as breeswish ...
save life, drop IE8!

**Comment by jeromelecomte-okta** [comment]
IE8 as a standalone browser is dead, no disagreement here. But the issue is embedded browsers such as the one used by Microsoft Office 365 which behaves like IE6. When you ship JS code that has to run in that environment (e.g. authenticating your user against an identify cloud), Webpack is no longer a viable option, which is unfortunate.

**Comment by jdalton** [comment]
I believe the HTML engine in Office is equiv to at least IE9+.

**Comment by jeromelecomte-okta** [comment]
Interesting, what's your source?

**Comment by jdalton** [comment]
@jeromelecomte-okta I made an office agave probably ~2years ago and I seemed to remember it being IE9+. I also did a quick search and turned up [this](http://blog.randomdust.com/2015/12/check-users-browser-within-office-365-add-in/). It's also the min bar noted [here](https://github.com/OfficeDev/Office-Add-in-JavaScript-WebApiService#prerequisites).

**Comment by inferpse** [comment]
@jeromelecomte-okta webpack is very flexible and it’s possible to create a plugin which will “downgrade” webpack2+ generated source to support ES3 environment.

Of course it is a dirty and unreliable way which could break in any next release
but it’s possible and here’s the PoC plugin which resolves that:
https://github.com/inferpse/es3-harmony-webpack-plugin

**Comment by jeromelecomte-okta** [comment]
Thanks for the info

**Comment by jeromelecomte-okta** [comment]
@jdalton for reference, this is the source I had found hinting Office 365 embedded browser was closer to IE7 than IE9: https://weblog.west-wind.com/posts/2011/May/21/Web-Browser-Control-Specifying-the-IE-Version

**Comment by jdalton** [comment]
No worries, it's not IE7.

**Comment by pauldwaite** [comment]
@jdalton 

> I believe no matter where you are there's access to other browsers.

Lots of users don’t get to choose what browser they run. I’m currently working on a project for internal UK government users, all of whom get a locked down computer to work on, running IE 8.

In addition, many of them don’t know what a browser is, let alone how to upgrade one.

**Comment by csvan** [comment]
That is true, but perpetually supporting dead browsers and environments is detrimental to the overall progress of both technology and society. Institutions need to understand that maintaining their IT infrastructure matters.

**Comment by pauldwaite** [comment]
@csvan sure. I don’t think anyone’s saying dead browsers should be perpetually supported. I think we’re saying IE 8 is not dead yet.

**Comment by csvan** [comment]
IE8 is dead because it is officially (way) past its EOL, no longer maintained by anyone, and no longer a viable option for new development - not because there are pockets of the market who still use it.

**Comment by pauldwaite** [comment]
@csvan it’s past EOL and no longer maintained, sure. When I say it’s not dead, I mean it’s still in use enough to warrant consideration by projects that build for the web.

And the outcome of that consideration can totally be “We don’t support it.” I personally find that disappointing from a tool that transpiles web code to work on older browsers, but it’s a valid position.

**Comment by jdalton** [comment]
> When I say it’s not dead, I mean it’s still in use enough to warrant consideration by projects that build for the web.

There isn't enough use to warrant consideration for projects for the web. This is why the JavaScript ecosystem as a whole has moved on. If you're disappointed you can always use older versions of all the tools and libraries in the ecosystem but it is unreasonable to demand support for it at this point.

**Comment by Justineo** [comment]
A little off the topic: I want to point out that users are not always able to choose the browsers they use. Like here in China, many students/teachers/doctors/…, they use computers provided by their schools/hospitals/governments/…, which might receive a system image recovery everyday to prevent malware/virus/…, especially for poverty areas, where many schools are built with donation so they have rather limited resouces to replace computers/OS. Many of these people *have to* use outdated browsers like IE8, even down to IE6. So I wouldn't put supporting legacy browsers as “irresponsible”. China's largest search engine still supports IE8, just to take care of those people, letting them have a chance to connect to the outside world. The good news is the market share of these legacy browsers are continiously going down, which helps Chinese developers to move on. I'm not saying webpack should continue to support legacy browsers forever, but please understand there might be a reason behind every developer who has to do so at the moment.

**Comment by haishengwu-okta** [comment]
just in case people search google and find this thread, this is how I solved at webpack 3 by updating the babel config.

```
presets: [
    ['es2015', {'loose': true} ]
],
plugins: [
    'transform-es3-member-expression-literals',
    'transform-es3-property-literals'
]
```
